app/webpack/users/edit/ducks/relationships.js

```diff
@@ -0,0 +1,332 @@
+import inatjs from "inaturalistjs";
+
+import { fetchUserSettings } from "./user_settings";
+
+const SET_RELATIONSHIPS = "user/edit/SET_RELATIONSHIPS";
+const SET_FILTERED_RELATIONSHIPS = "user/edit/SET_FILTERED_RELATIONSHIPS";
+const SET_FILTERS = "user/edit/SET_FILTERS";
+const SET_RELATIONSHIP_TO_DELETE = "user/edit/SET_RELATIONSHIP_TO_DELETE";
+const SET_USER_AUTOCOMPLETE = "user/edit/SET_USER_AUTOCOMPLETE";
+const SET_BLOCKED_USERS = "user/edit/SET_BLOCKED_USERS";
+const SET_MUTED_USERS = "user/edit/SET_MUTED_USERS";
+const SET_PAGE = "user/edit/SET_PAGE";
+
+export default function reducer( state = {
+  filters: { name: null, following: "all", trusted: "all" },
+  blockedUsers: [],
+  mutedUsers: [],
+  page: 1
+}, action ) {
+  switch ( action.type ) {
+    case SET_RELATIONSHIPS:
+      return { ...state, relationships: action.relationships };
+    case SET_FILTERED_RELATIONSHIPS:
+      return { ...state, filteredRelationships: action.filteredRelationships };
+    case SET_FILTERS:
+      return { ...state, filters: action.filters };
+    case SET_RELATIONSHIP_TO_DELETE:
+      return { ...state, id: action.id };
+    case SET_USER_AUTOCOMPLETE:
+      return { ...state, users: action.users };
+    case SET_BLOCKED_USERS:
+      return { ...state, blockedUsers: action.blockedUsers };
+    case SET_MUTED_USERS:
+      return { ...state, mutedUsers: action.mutedUsers };
+    case SET_PAGE:
+      return { ...state, page: action.page };
+    default:
+  }
+  return state;
+}
+
+export function setRelationships( relationships ) {
+  return {
+    type: SET_RELATIONSHIPS,
+    relationships
+  };
+}
+
+export function setFilteredRelationships( filteredRelationships ) {
+  return {
+    type: SET_FILTERED_RELATIONSHIPS,
+    filteredRelationships
+  };
+}
+
+export function setFilters( filters ) {
+  return {
+    type: SET_FILTERS,
+    filters
+  };
+}
+
+export function setRelationshipToDelete( id ) {
+  return {
+    type: SET_RELATIONSHIP_TO_DELETE,
+    id
+  };
+}
+
+export function setUserAutocomplete( users ) {
+  return {
+    type: SET_USER_AUTOCOMPLETE,
+    users
+  };
+}
+
+export function setBlockedUsers( blockedUsers ) {
+  return {
+    type: SET_BLOCKED_USERS,
+    blockedUsers
+  };
+}
+
+export function setMutedUsers( mutedUsers ) {
+  return {
+    type: SET_MUTED_USERS,
+    mutedUsers
+  };
+}
+
+export function setPage( page ) {
+  return {
+    type: SET_PAGE,
+    page
+  };
+}
+
+export function filterRelationships( ) {
```

Comment:
Not sure where to put this so I'll put it here: are you only filtering here in the client and not making a new API request? I think that's what's happening, which means if you have a hundred relationships, the filters you apply will only get applied to the 10 relationships on the first page. If I'm misreading this, my apologies. Local dev env is hosed atm so I can't quite test as thoroughly as I'd like.

Reply:
I'm filtering in the client, and it is applied to all relationships. I'm basically fetching all relationships, then storing the result in the `relationships` state, and using that list as the base of truth to apply filters, which avoids only filtering the 10 displayed relationships.

This code all felt awkward to write though, so I'm not surprised it's unclear to read. Open to suggestions and also curious if it would make more sense for the relationships endpoint to handle the sorting and filtering?

Reply:
Hah, I probably should have read this before writing my other reply. FWIW, I don't think [fetchRelationships](https://github.com/inaturalist/inaturalist/pull/2865/files#diff-1f1c2c9fea07980531091f52d8b3d905b2db0fbf026a63ed3ffdf20091c33b31R210) is fetching all relationships, just the first page. And yes, I think the search / filter / page logic should be handled via the API. FWIW, the API is passing things through to the Rails app, which ends up getting handled in [`FriendshipsController#index`](https://github.com/inaturalist/inaturalist/blob/main/app/controllers/friendships_controller.rb#L8).

Reply:
I think I have everything working from the API except for search. Does [this line](https://github.com/inaturalist/iNaturalistAPI/blob/2dc16cb6357f15555c038343fccdf707be3361a6/lib/controllers/v1/relationships_controller.js#L30) mean that I should be able to request something like `/v1/relationships?q=kueda` to search user logins? (When I try adding that query I'm getting a 500 error).

Reply:
Sorry about that, should be fixed in https://github.com/inaturalist/iNaturalistAPI/commit/05c6582a1d569eb1a9a0d098bfe9ffa91bffd860

Reply:
Finished this -- search is working now.